File: 본문/java/Code_3_07_3.java
import java.util.*;

class Code_3_07_3 {
	public static void main(String[] args) {
		// 입력
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int W = sc.nextInt();
		int[] w = new int[N + 1];
		int[] v = new int[N + 1];
		for (int i = 1; i <= N; i++) {
			w[i] = sc.nextInt();
			v[i] = sc.nextInt();
		}
		
		// 배열 초기화
		long INF = 1000000000000000000L;
		long[][] dp = new long[N + 1][W + 1];
		dp[0][0] = 0;
		for (int i = 1; i <= W; i++) {
			dp[0][i] = -INF;
		}
		
		// 동적계획법
		for (int i = 1; i <= N; i++) {
			for (int j = 0; j <= W; j++) {
				if (j < w[i]) {
					// j < w[i]라면, 방법A를 선택할 수 밖에 없음
					dp[i][j] = dp[i - 1][j];
				}
				else {
					// j >= w[i]라면, 방법A/방법B 중에 아무 것이나 선택할 수 있음
					dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - w[i]] + v[i]);
				}
			}
		}
		
		// 답을 계산하고 출력
		long answer = 0;
		for (int i = 0; i <= W; i++) {
			answer = Math.max(answer, dp[N][i]);
		}
		System.out.println(answer);
	}
}

File: 본문/python/Code_3_07_3.py
# 입력
N, W = map(int, input().split())
w = [ None ] * N
v = [ None ] * N
for i in range(N):
	w[i], v[i] = map(int, input().split())

# 배열 초기화
INF = 10 ** 18
dp = [ [ None ] * (W + 1) for i in range(N + 1) ]
dp[0][0] = 0
for i in range(1, W + 1):
	dp[0][i] = -INF

# 동적계획법
for i in range(1, N + 1):
	for j in range(0, W + 1):
		if j < w[i - 1]:
			# j < w[i-1]라면, 방법A로 선택
			dp[i][j] = dp[i - 1][j]
		else:
			# j >= w[i-1]라면 방법A와 방법B 중에 큰 것으로 선택
			dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - w[i - 1]] + v[i - 1])

# 답을 계산하고 출력
answer = max(dp[N])
print(answer)
